File: Homework/Hang_Nguyen/lesson_1/range-sum-query-immutable.py
class NumArray:

    def __init__(self, nums):
        self.sumLst = []
        s = 0
        for num in nums:
            s += num
            self.sumLst.append(s)

    
    def sumRange(self, left, right):
        if left > 0:
            return self.sumLst[right] - self.sumLst[left-1]
        return self.sumLst[right]


# TEST CASE
numArray = NumArray([-2, 0, 3, -5, 2, -1])

t = numArray.sumRange(0, 2)
print("TEST 1 is " + str(t == 1))

t = numArray.sumRange(2, 5)
print("TEST 2 is " + str(t == -1))

t = numArray.sumRange(0, 5)
print("TEST 3 is " + str(t == -3))

File: Homework/Nguyen_Anh_Duc/range-sum-query-immutable.go
package main

import "fmt"

type NumArray struct {
	nums []int
}

func Constructor(nums []int) NumArray {
	a := NumArray{}
	a.nums = nums
	for i, num := range nums {
		if i != 0 {
			a.nums[i] = num + a.nums[i-1]
		}
	}
	return a
}

func (this *NumArray) SumRange(left int, right int) int {
	if left == 0 {
		return this.nums[right]
	}
	return this.nums[right] - this.nums[left-1]
}

func main() {
	n := Constructor([]int{-2, 0, 3, -5, 2, -1})
	fmt.Println("0,2", n.SumRange(0, 2))
	fmt.Println("2,5", n.SumRange(2, 5))
	fmt.Println("0,5", n.SumRange(0, 5))
}
